proof/crefine/RISCV64/SchedContext_C.thy

```diff
@@ -584,11 +584,251 @@ lemma schedContext_updateConsumed_ccorres:
+           apply (rule ccorres_return_Skip)
+          apply (wpsimp wp: hoare_drop_imps)
+         apply (wpsimp wp: tcbReleaseRemove_invs')
+        apply (rule ccorres_return_Skip)
+       apply ceqv
+      apply (rule_tac r'=dc and xf'=xfdc in ccorres_split_nothrow_novcg)
+          apply (rule ccorres_cond_both'[where Q="\<top>" and Q'=\<top>])
+            apply fastforce
+           apply (ctac (no_vcg) add: refill_new_ccorres)
+          apply clarsimp
+          apply (rule ccorres_rhs_assoc)+
+          apply (ctac (no_vcg) add: sc_active_ccorres, rename_tac active)
+           apply csymbr
+           apply clarsimp
+           apply (rule_tac P="to_bool active" in ccorres_cases)
+            apply ccorres_rewrite
+            apply (rule ccorres_cond_seq)
+            apply (rule ccorres_cond_true)
+            apply (rule_tac xf'=ret__int_'
+                        and val="from_bool (scTCB sc \<noteq> None)"
+                        and R="obj_at' (\<lambda>sc'. scTCB sc' = scTCB sc) scPtr and valid_objs'"
+                         in ccorres_symb_exec_r_known_rv[where R'="UNIV"])
+               apply (rule conseqPre, vcg)
+               apply normalise_obj_at'
+               apply (frule (1) sc_ko_at_valid_objs_valid_sc')
+               apply (frule (1) obj_at_cslift_sc)
+               apply (clarsimp simp: option_to_ctcb_ptr_def valid_sched_context'_def)
+               apply (case_tac "scTCB sc"; clarsimp)
+               apply (fastforce dest: tcb_at_not_NULL)
+              apply ceqv
+             apply (rule ccorres_cond_seq)
+             apply (rule ccorres_cond_both'[where Q="\<top>" and Q'=\<top>])
+               apply fastforce
+              apply clarsimp
+              apply (drule Some_to_the)
+              apply (clarsimp simp: option_to_ctcb_ptr_def split: option.splits)
+              apply (rule ccorres_rhs_assoc)+
+              apply (ctac (no_vcg) add: isRunnable_ccorres)
+               apply csymbr
+               apply (rule ccorres_cond_both'[where Q="\<top>" and Q'=\<top>])
+                 apply (clarsimp simp: to_bool_def)
+                apply (ctac (no_vcg) add: refill_update_ccorres)
+               apply (ctac (no_vcg) add: refill_new_ccorres)
+              apply wpsimp
+             (* target sc is not associated with a thread, so we perform refill_new *)
+             apply ccorres_rewrite
+             apply (rule ccorres_cond_false)
+             apply (ctac add: refill_new_ccorres)
+            apply vcg
+           (* target sc is not active, so we perform refill_new *)
+           apply clarsimp
+           apply ccorres_rewrite
+           apply (rule ccorres_cond_seq)
+           apply (rule ccorres_cond_false)
+           apply ccorres_rewrite
+           apply (rule ccorres_cond_seq)
+           apply (rule ccorres_cond_false)
+           apply ccorres_rewrite
+           apply (rule ccorres_cond_false)
+           apply (ctac (no_vcg) add: refill_new_ccorres)
```

Comment:
little bit confused here; I'm only seeing one ccorres_cases above, we're in the not active case, but you're clearing out a bunch of branches (and they're not clearing up by themselves either, it's quite forceful) - can you say what's happening here?

Reply:
Yes, in the case that the sched context is not active, we will perform `refill_new`. But the Simpl will just use the variable `ret_int_'` to keep track of the truth value of the guards that it uses. So when the sc is not active, it assigns `0` to `ret_int_'`, and then it will take the false branch of every conditional until we eventually get to `refill_new`

Reply:
right, but that's why I'm surprised... it sounds like simp, ccorres_rewrite, simp, ccorres_rewrite, etc, since it never takes the branches, but you're doing explicit ccorres_rewrite_cond_false which defers showing that it's false until later

Reply:
Sorry, is `ccorres_rewrite_cond_false` a rule? If so, I can't find it. I'm not sure how else this can be handled. I really do not like the way that the CParser handles these sorts of `if` statements and uses local variables to determine the guards. If there is a nicer way, please let me know.

Reply:
Apologies, that's a typo by me, I meant the `ccorres_cond_false` rule you're using. What I'm trying to get at is that if we know the return for the first ret_int, we can csymbr that, and we end up with something in the next IF branch condition. If we simp that, it should boil down to the empty set, which we can discard with ccorres_rewrite. So my mind is seeing some kind of a `csymbr, simp, ccorres_rewrite` that will clear up the branches. It's ok to do it the way you did, but it is surprising to be deferring these IF conditions 'til later when they seem obvious in the current context.